### ISSUE TITLE
Feature Request: Raspberry pi as USDX-2TV box + 12 mobile phones as microphones

### ISSUE THREAD

**Issue by segalion** [UltraStar-Deluxe/USDX#712]
Please, could you begin to think in an integration with mumble in order to get a quick configuration to automagically configure users from mobile?

Today, we "easily" can run 6 players that just can install a mumble client on their mobile, and connect to rpi by wifi and act as a virtual microphone. This is with mumble-server, 12 groups (one per player), a (silent and no-screen) local mumble client per slot, and an pulseaudio configurarion to create 12 virtual microphones from each local mumble client.

The RPI acts as a wifi-hotspot, and can manage up to 2 HDMI 1080 TV screens.
People connect to ULTRASTAR-DX wifi, and execute the a mumble client to connect to one of all available groups. Even more than one players can connect to same mumble group to use same USDX microphone.

The problem is that you need to make a lot of manual configurations, and finally manually make GUI configurations in USDX to match every virtual microphone with every player.

Could be great that USDX incorporate all this stuff to start everything and even autodetect usernames of people (from mumble) connecting to slots (groups) in mumble-server. Could be great that USDX incorporates the mumble client code inside.

Thanks

**Comment by s09bQ5** [comment]
Wouldn't it make more sense to have a separate program/script that sets up Mumble and PulseAudio and then changes the config.ini file of USDX to have players assigned the correct devices and names?

**Comment by segalion** [comment]
> Wouldn't it make more sense to have a separate program/script that sets up Mumble and PulseAudio and then changes the config.ini file of USDX to have players assigned the correct devices and names?

Yes, but I need to rewrite .ini and restart USDX, at every player change... Could be great a command line to force USDX to re-read config file...

This is the first aproach.

**Comment by barbeque-squared** [comment]
I've used this setup in the past (it works great). I also ran into the "update player names" issue: it seems there are some Lua API bits for it, but something tells me they only work for Party mode or something. In the end I just (ab)used the Lua API to detect which screen we were on, and if it's on the sing screen, get the player names from Mumble (python-zeroc-ice) and hard overlay them with I think it was osd_cat (or xosd).

Not ideal, and especially if the player name bit of the API is actually already in place, just not working for regular singing, I think that is something we can/should fix. I'm assuming you just have some generic avatar set for each player, but want the correct name for highscore reasons?

The rest though: it's going to be very specific for each setup. Not everyone will be using pulseaudio, and it's only a matter of time before people want "that one feature" in their Mumble setup. And USDX would need to decide which player name(s) to assign if there's multiple people in one channel -- that's not something USDX should concern itself with imo.

I've never needed to assign my virtual microphones through the GUI though. USDX detects them with a name I can set myself -- this can also be done for _real_ microphones -- so all my mumble ones are just `mumble1`, `mumble2` etc. I never bothered but if needed I could just generate that entire .ini section.

@segalion would being able to update the USDX player names from external programs be a good first step for this?

**Comment by s09bQ5** [comment]
Sounds a bit like a job for a D-Bus interface. But that would be a lot more work than a key binding to reload config.ini.

**Comment by segalion** [comment]
> The rest though: it's going to be very specific for each setup. Not everyone will be using pulseaudio, and it's only a matter of time before people want "that one feature" in their Mumble setup. And USDX would need to decide which player name(s) to assign if there's multiple people in one channel -- that's not something USDX should concern itself with imo.

I am talking about one specific platform that seems perfect for this game. Obviously this solution is based in pulseaudio, so only available in linux platforms.

It is perfect to be able to take a raspberry anywhere (a cheap little box the size of a mouse), with the USDX and all the songs, connect it to any television and tell a lot of people that they only have to install an app, connect to the wifi ULTRASTAR-DX run the app, enter the IP address (a pity that android doesn't implement bonjour) and each connect to a channel and start singing.

For me, being able to take Karaoke like this anywhere is priceless. It is a configuration that brings so much value that I think that it should be given a little attention from development, simply to improve all these aspects of implementation and usability. This was a bit the objective that I intended with this "Feature Request"

Also, although it is not very friendly, I can control the USDX GUI from my mobile through VNC (and I avoid having to carry a keyboard)

Obviously, the ideal would be to design an app for Android and IOS that integrates perfectly with USDX, but I understand that this would be a very Long Term Design. 

But for now, I think we could get some kind of integration with this config.

**Comment by dgruss** [comment]
not 12 microphones yet, but other than that https://github.com/dgruss/SmartMicrophone addresses exactly this issue

**Comment by barbeque-squared** [comment]
Re: 12 microphones: I ran a proof-of-concept 12 players / 2 screens at an event in October this year. It mostly worked (there were some visual bugs), but it also takes like 1000+ lines of code to do this. The main issue is having to add a check for "are we running 12 players _and_ on 2 screens" in entirely too many (often completely irrelevant/illogical) places to be anywhere remotely maintainable.

This isn't anything specific to 12 players, all the existing modes (1, 2, 3, 4 or 6 players times 1 or 2 screens) also have this issue.

Bottom line: 12 players isn't happening anytime soon, at the very least not before ~~everything~~ most of what's wrong with the existing modes has been addressed.

**Comment by dgruss** [comment]
12+ players is working very well with #1186. and #1186 actually leads to a net reduction of 3000 loc because it is just a clean refactor, supporting both single screen and dual screen

@segalion if you like you can try it out here: #1186
